Add support for new Event Meters

Bark-fa opened the pull request:
Hello! This PR adds support for the new event meters by stripe for tracking usage in usage-based subscriptions, this is my first time contributing to cashier, and I've made every effort to maintain the style and naming conventions, and everything is fully typed, please let me know if I missed something either in the code itself or the tests.

Thank you so much for your time and effort.

Bark-fa committed:
commit before writing tests

Bark-fa committed:
Add support for new Event Meters

Bark-fa commented:
I see the CI has failed but for some reason, I cannot see why, it shows a generic "blocked" message, is it some geo-restriction?

Bark-fa committed:
Updated styles and removed unused imports

Bark-fa committed:
Updated styles and removed unused imports

Bark-fa committed:
Additional style fixes

Bark-fa commented:
I managed to access it with a VPN, all style issues have been fixed

driesvints commented:
Can you remove all of your idea settings from this PR? Thanks!

Bark-fa commented:
Ah yes, sorry about that I thought they'd be in the .gitignore!

Bark-fa committed:
removed idea files

Bark-fa commented:
done @driesvints

Bark-fa commented:
@driesvints Do you need me to delete them from the previous commit history too?

driesvints commented:
@Bark-fa no that's okay. This PR will get squashed anyway.

Bark-fa commented:
Hello @driesvints, I'm reaching out because I noticed the PR is still in draft, I know you must be very busy, so if the PR needs any more work or if I missed anything in the docs please let me know and I'll gladly take it off your hands, I've got plenty of time this week.

Thanks!

driesvints commented on src/Subscription.php:
```diff
@@ -552,6 +554,25 @@ public function reportUsage($quantity = 1, $timestamp = null, $price = null)
+     * @param  string  $meter
+     * @param  int  $quantity
+     * @param  string|null  $price
+     * @return MeterEvent
+     *
+     * @throws ApiErrorException
```
```suggestion
     * @return \Stripe\MeterEvent
```

driesvints commented on src/Subscription.php:
```diff
@@ -565,6 +586,21 @@ public function reportUsageFor($price, $quantity = 1, $timestamp = null)
+     * @param  string  $price
+     * @param  int  $quantity
+     * @param  string  $eventName
+     * @return MeterEvent
+     *
+     * @throws ApiErrorException
```
```suggestion
     * @return \Stripe\MeterEvent
```

driesvints commented on src/Subscription.php:
```diff
@@ -581,6 +617,25 @@ public function usageRecords(array $options = [], $price = null)
+     *
+     * @param  string  $meterId
+     * @param  array  $options
+     * @param  null  $price
+     * @return Collection
+     *
+     * @throws ApiErrorException
```
```suggestion
     * @param  string|null  $price
     * @return \Illuminate\Support\Collection
```

driesvints commented on src/SubscriptionItem.php:
```diff
@@ -220,6 +222,26 @@ public function reportUsage($quantity = 1, $timestamp = null)
+     *
+     * @param  string  $meter
+     * @param  int  $quantity
+     * @return MeterEvent
+     *
+     * @throws ApiErrorException
```
```suggestion
     * @return \Stripe\MeterEvent
```

driesvints commented on src/SubscriptionItem.php:
```diff
@@ -233,6 +255,49 @@ public function usageRecords($options = [])
+     *
+     * @param  array|null  $params
+     * @param  array|null  $opts
+     * @return Collection
+     *
+     * @throws ApiErrorException
```
```suggestion
     * @return \Illuminate\Support\Collection
```

driesvints commented on src/SubscriptionItem.php:
```diff
@@ -233,6 +255,49 @@ public function usageRecords($options = [])
+     * @param  string  $meterId
+     * @param  array|null  $params
+     * @param  array|null  $opts
+     * @return Collection
+     *
+     * @throws ApiErrorException
```
```suggestion
     * @return \Illuminate\Support\Collection
```

driesvints committed:
Update src/Subscription.php

driesvints committed:
Update src/Subscription.php

driesvints committed:
Update src/Subscription.php

driesvints committed:
Update src/SubscriptionItem.php

driesvints committed:
Update src/SubscriptionItem.php

driesvints committed:
Update src/SubscriptionItem.php

driesvints committed:
wip

driesvints committed:
wip

driesvints committed:
wip

driesvints committed:
wip

driesvints committed:
wip

driesvints commented:
@Bark-fa sorry, needed a bit to have a thorough look here. I completely refactored all of this to be on a new trait on the billable. It's not needed for usage based billing to live closely to a subscription or subscription item. I've revised the API's and simplified things. 

The downside of this is that the written tests aren't applicable anymore. Could you rewrite those into a new `UsageBasedBillingTest.php` file? Then I believe this PR will be ready for review.

Bark-fa committed:
Refactor tests

Bark-fa commented:
Done, I refactored the new tests into `UsageBasedBillingTest.php` and removed them from the old file.

Let me know if anything else comes up, thank you so much!

Bark-fa commented:
And regarding the new trait, I agree it's a better approach, I didn't allow myself the freedom to create new files though so I just wrote the functions where the old (now deprecated) code was 😅

driesvints committed:
wip

driesvints committed:
wip

driesvints commented:
Done. Thanks a bunch for this one @Bark-fa. Would also highly appreciate a PR to the docs!

Bark-fa commented:
Hello @driesvints, you're very much welcome, I just submitted a [PR for the docs](https://github.com/laravel/docs/pull/9814)

taylorotwell committed:
Update ManagesUsageBilling.php

driesvints commented:
@Bark-fa did you try these tests? Because they're failing atm.

Bark-fa commented:
Hello @driesvints, what tests specifically? I tried all the tests I wrote, and the old ones as well, and all ran with success, which ones are failing?

driesvints commented:
These: https://github.com/laravel/cashier-stripe/actions/runs/10444767851/job/28919710065

Bark-fa commented:
I just checked, I think the problem was caused by this line:
` $startTime = $options['start_time'] ?? $this->created_at->timestamp;`

This used to be in the Subscription class, so the logic was that if the user did not provide a start time, we'd set it to the date the subscription was created, but when it was moved to a trait, $this references the User model, this should fix it:
`        $startTime = $options['start_time'] ?? 1;`

I pushed it to my fork

driesvints commented:
Can you send in a PR? Thanks!

Bark-fa commented:
Of course, on it
